File: WebAPI/Controllers/BookingController.cs
namespace WebAPI.Controllers;

using BusinessLayer.Contracts;
using BusinessLayer.Models;
using BusinessLayer.Services.Interfaces;
using Microsoft.AspNetCore.Mvc;

[Route("api/booking")]
[ApiController]
public class BookingController : ControllerBase
{
    private readonly IBookingService _bookingService;
    private readonly ISpaceService _spaceService;

    public BookingController(IBookingService bookingService, ISpaceService spaceService)
    {
        _bookingService = bookingService;
        _spaceService = spaceService;
    }

    [HttpGet("all")]
    public ActionResult<CollectionResponse<SpaceBookingsDTO>> GetBookings()
    {
        var response = _spaceService.GetAllSpacesWithBookings();

        if (response.HasError())
        {
            return BadRequest(response.Error);
        }

        return Ok(response);
    }

    [HttpGet("bookingInfo")]
    public ActionResult<BookingInfoResponse> GetBookingInformation([FromQuery] BookingRequest request)
    {
        var response = _bookingService.GetBookingInfo(request);

        if (response.HasError())
        {
            return BadRequest(response.Error);
        }

        return Ok(response);
    }

    [HttpPost("bookingInfo/book")]
    public async Task<ActionResult<BaseResponse>> Book([FromBody] BookingDTO booking)
    {
        var response = await _bookingService.AddBooking(booking);

        if (response.HasError())
        {
            return BadRequest(response.Error);
        }

        return Ok(response);
    }

    [HttpGet("all-bookings")]
    public ActionResult<CollectionResponse<BookingDetailsDTO>> GetAllBookings()
    {
        var response = _bookingService.GetAllBookings();

        if (response.HasError())
        {
            return BadRequest(response);
        }

        return Ok(response);
    }

    [HttpPatch("delete/{id}")]
    public async Task<ActionResult<BaseResponse>> DeleteBooking([FromBody] Guid id)
    {
        var response = await _bookingService.DeleteBooking(id);

        if (response.HasError())
        {
            return BadRequest(response.Error);
        }

        return Ok(response);
    }
}

File: Front-end/WebLayer/Controllers/BookingController.js
import { BookingService } from "../../BusinessLayer/Services/BookingService.js";
import { SpaceMap } from "../../BusinessLayer/Mappers/SpaceMap.js";
import { BookingMapper } from "../../BusinessLayer/Mappers/BookingMapper.js";
import { displayPrompt } from "../utilities/Prompt.js";


export class BookingController
{
    constructor()
    {
        this.bookingService = new BookingService();
        this.spaceMapper = new SpaceMap();
        this.bookingMapper = new BookingMapper();
    };
    
    async GetAll()
    {
        let response = await this.bookingService.getAll();
        if(response.hasError)
        {
            await displayPrompt('.prompt-save', response.errorMessage, false);
            return;
        };
        let bookings = response.data.map(spaceData=>this.spaceMapper.MapToSpace(spaceData,true));
        return bookings;
    };

    async GetBookingInfo()
    {
        let response = await this.bookingService.getBookingInfo();
        if(response.error.hasError)
        {
            await displayPrompt('.prompt-save', response.error.errorMessage, false);
            return;
        };
        return response;
    };

    async Book(booking)
    {
        let bookingDTO = this.bookingMapper.MapToBookingDTO(booking);
        let response = await this.bookingService.book(bookingDTO);
        if(response.error.hasError)
        {
            await displayPrompt('.prompt-save', response.error.errorMessage, false);
            return;
        };
        return response;
    };

    async GetAllBookings()
    {
        let response = await this.bookingService.getAllBookings();
        if(response.error.hasError)
        {
            await displayPrompt('.prompt-save', response.error.errorMessage, false);
            return;
        };
        let bookings = response.data.map(bookingData=>this.bookingMapper.MapToBooking(bookingData));
        return bookings;
    };

    async DeleteBooking(id)
    {
        let response = await this.bookingService.deleteBooking(id);
        if(response.error.hasError)
        {
            await displayPrompt('.prompt-save', response.error.errorMessage, false);
            return;
        };
        await displayPrompt('.prompt-save', 'Успешно изтрихте резервацията.', false);
        return;
    };
    
}